pinot-segment-local/src/main/java/org/apache/pinot/segment/local/recordtransformer/ComplexTypeTransformer.java

```diff
@@ -0,0 +1,252 @@
+      _unnestFields = tableConfig.getIngestionConfig().getComplexTypeConfig().getUnnestConfig() != null ? tableConfig
+          .getIngestionConfig().getComplexTypeConfig().getUnnestConfig() : new ArrayList<>();
+    } else {
+      _unnestFields = new ArrayList<>();
+    }
+  }
+
+  @VisibleForTesting
+  public ComplexTypeTransformer(List<String> unnestCollections) {
+    _unnestFields = new ArrayList<>(unnestCollections);
+    Collections.sort(_unnestFields);
+  }
+
+  public static boolean isComplexTypeHandlingEnabled(TableConfig tableConfig) {
+    return tableConfig.getIngestionConfig() != null && tableConfig.getIngestionConfig().getComplexTypeConfig() != null;
+  }
+
+  @Nullable
+  @Override
+  public GenericRow transform(GenericRow record) {
+    flattenMap(record, new ArrayList<>(record.getFieldToValueMap().keySet()));
+    for (String collection : _unnestFields) {
+      unnestCollection(record, collection);
+    }
+    return record;
+  }
+
+  private GenericRow unnestCollection(GenericRow record, String column) {
+    if (record.getValue(GenericRow.MULTIPLE_RECORDS_KEY) == null) {
+      List<GenericRow> list = new ArrayList<>();
+      unnestCollection(record, column, list);
+      record.putValue(GenericRow.MULTIPLE_RECORDS_KEY, list);
+    } else {
+      Collection<GenericRow> records = (Collection) record.getValue(GenericRow.MULTIPLE_RECORDS_KEY);
+      List<GenericRow> list = new ArrayList<>();
+      for (GenericRow innerRecord : records) {
+        unnestCollection(innerRecord, column, list);
+      }
+      record.putValue(GenericRow.MULTIPLE_RECORDS_KEY, list);
+    }
+    return record;
+  }
+
+  private void unnestCollection(GenericRow record, String column, List<GenericRow> list) {
+    Object value = record.removeValue(column);
+    if (value == null) {
+      // use the record itself
+      list.add(record);
+    } else if (value instanceof Collection) {
+      if (((Collection) value).isEmpty()) {
+        // use the record itself
+        list.add(record);
+      } else {
+        for (Object obj : (Collection) value) {
+          GenericRow copy = flattenCollectionItem(record, obj, column);
+          list.add(copy);
+        }
+      }
+    } else if (isArray(value)) {
+      if (((Object[]) value).length == 0) {
+        // use the record itself
+        list.add(record);
+      } else {
+        for (Object obj : (Object[]) value) {
+          GenericRow copy = flattenCollectionItem(record, obj, column);
+          list.add(copy);
+        }
+      }
+    }
+  }
+
+  private GenericRow flattenCollectionItem(GenericRow record, Object obj, String column) {
+    GenericRow copy = record.copy();
+    if (obj instanceof Map) {
+      Map<String, Object> map = (Map<String, Object>) obj;
+      for (Map.Entry<String, Object> entry : map.entrySet()) {
+        String flattenName = concat(column, entry.getKey());
+        copy.putValue(flattenName, entry.getValue());
+      }
+    } else {
```

Comment:
Should we also handle multi-dimensional array here? You may pass in the list, and make it recursive

Reply:
No, it's not needed, since the caller will take care of it by iterating the `unnestFields` after sort.